src/Components/Server/src/Circuits/ServerComponentDeserializer.cs

```diff
@@ -147,6 +147,12 @@ public bool TryDeserializeComponentDescriptorCollection(string serializedCompone
         return true;
     }
 
+    public ComponentDescriptor DeserializeServerComponentDescriptor(ServerComponentMarker record)
+    {
+        var (descriptor, _) = DeserializeServerComponent(record);
+        return descriptor;
+    }
+
```

Comment:
The requirement to deserialize descriptors in the order that they were rendered (and without skipping any) doesn't map perfectly to this new rendering model. Descriptors could be skipped due to being replaced or removed on the page before getting activated as interactive components. Furthermore, some situations warrant handling descriptors in the wrong order. For example, components added during a streaming update don't become interactive components until the streaming request completes. However, already-interactive components should receive parameter updates immediately.

Fundamentally, we are trying to make sure that descriptors are legitimate and can't get used more than once. Data protection should guarantee the legitimacy of the descriptor, so we would just need to find an alternative way to guarantee that they don't get reused.

Reply:
> we would just need to find an alternative way to guarantee that they don't get reused.

Would it be enough just to have a `HashSet<int>` on `RemoteRenderer` tracking the ones we've consumed? If we know DataProtection ensures the client can't make up new ID numbers on its own, then the total size would be bounded.

Reply:
Something like that should work. The other thing to consider is that the sequence ID will start over at 0 for each new HTTP response, so we'll have to pair it with the invocation ID (which is unique to the HTTP response) as well.

If we're worried about the hash set becoming too large, we could make this slightly better by adding a rule where after we deserialize a descriptor with an unseen invocation ID, all previously seen invocation IDs become invalid. So, you have two hash sets, one that stores consumed invocation IDs, and one that stores consumed sequence IDs for the current invocation ID.